stdlib/LinearAlgebra/src/lq.jl

```diff
@@ -303,36 +307,33 @@ _rightappdimmismatch(rowsorcols) =
         "(the factorization's originating matrix's number of rows)")))
 
 
-function (\)(A::LQ{TA}, b::StridedVector{Tb}) where {TA,Tb}
-    S = promote_type(TA,Tb)
-    m = checksquare(A)
-    m == length(b) || throw(DimensionMismatch("left hand side has $m rows, but right hand side has length $(length(b))"))
-    AA = Factorization{S}(A)
-    x = ldiv!(AA, copy_oftype(b, S))
-    return x
-end
-function (\)(A::LQ{TA},B::StridedMatrix{TB}) where {TA,TB}
+function (\)(A::LQ{TA},B::StridedVecOrMat{TB}) where {TA,TB}
     S = promote_type(TA,TB)
-    m = checksquare(A)
-    m == size(B,1) || throw(DimensionMismatch("left hand side has $m rows, but right hand side has $(size(B,1)) rows"))
+    m, n = size(A)
+    m ≤ n || throw(DimensionMismatch("LQ solver does not support underdetermined systems (more rows than columns)"))
+    m == size(B,1) || throw(DimensionMismatch("Both inputs should have the same number of rows"))
     AA = Factorization{S}(A)
-    X = ldiv!(AA, copy_oftype(B, S))
-    return X
+    X = _zeros(S, B, n)
+    X[1:size(B, 1), :] = B
+    return ldiv!(AA, X)
 end
 # With a real lhs and complex rhs with the same precision, we can reinterpret
 # the complex rhs as a real rhs with twice the number of columns
 function (\)(F::LQ{T}, B::VecOrMat{Complex{T}}) where T<:BlasReal
     require_one_based_indexing(B)
-    c2r = reshape(copy(transpose(reinterpret(T, reshape(B, (1, length(B)))))), size(B, 1), 2*size(B, 2))
-    x = ldiv!(F, c2r)
-    return reshape(copy(reinterpret(Complex{T}, copy(transpose(reshape(x, div(length(x), 2), 2))))),
+    X = zeros(T, size(F,2), 2*size(B,2))
+    X[1:size(B,1), 1:size(B,2)] .= real.(B)
+    X[1:size(B,1), size(B,2)+1:size(X,2)] .= imag.(B)
+    ldiv!(F, X)
+    return reshape(copy(reinterpret(Complex{T}, copy(transpose(reshape(X, div(length(X), 2), 2))))),
                            isa(B, AbstractVector) ? (size(F,2),) : (size(F,2), size(B,2)))
 end
 
 
 function ldiv!(A::LQ{T}, B::StridedVecOrMat{T}) where T
-    lmul!(adjoint(A.Q), ldiv!(LowerTriangular(A.L),B))
-    return B
+    require_one_based_indexing(B)
+    ldiv!(LowerTriangular(A.L), view(B, 1:size(A,1), axes(B,2)))
+    return lmul!(adjoint(A.Q), B)
```

Comment:
This seems to error when I checked manually. BTW, are we testing `ldiv!` for non-square LQ? I could only find `\` tests.

Reply:
`\` calls `ldiv!`

Reply:
Yes, but `\` allocates a correctly sized vector/matrix `X` that has more rows than `B`. I guess my question is: can you do `ldiv!(lq(rand(2,5)), rand(2))`? It must return a vector of length 5, but I don't see that `ldiv!` allocates such a vector, or `append!`s some trailing zeros to the input `B`.

Reply:
I'm getting
```julia
julia> ldiv!(lq(rand(2,5)), rand(2))
ERROR: DimensionMismatch("for a left-sided multiplication, the first dimension of C, 2, must equal the second dimension of A, 5")
Stacktrace:
 [1] ormlq!(::Char, ::Char, ::Array{Float64,2}, ::Array{Float64,1}, ::Array{Float64,1}) at /Users/karrasch/Documents/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/lapack.jl:2715
 [2] lmul! at /Users/karrasch/Documents/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/lq.jl:198 [inlined]
 [3] ldiv!(::LQ{Float64,Array{Float64,2}}, ::Array{Float64,1}) at /Users/karrasch/Documents/julia/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/lq.jl:336
 [4] top-level scope at REPL[2]:1

julia> lq(rand(2,5)) \ rand(2)
5-element Array{Float64,1}:
 -0.21105452790704704
  0.41169392192670684
 -0.4022006494025896
  0.44360732676056863
  0.2111926086486643
```
with this PR. So we may need to append zeros to the rhs within `ldiv!`. Which leaves open what to do in the matrix rhs case: have a 3-arg `ldiv!` and solve columnwise?

Reply:
> I guess my question is: can you do `ldiv!(lq(rand(2,5)), rand(2))`? 

No, *you* need to append 3 rows of padding to the right-hand side so that it holds space for the solution.  It won't allocate, because the whole point of `ldiv!` is to work in-place.

For example:
```
julia> A = rand(2,5);

julia> A \ [3,4]
5-element Array{Float64,1}:
  1.798302830371654
  0.1864736948656177
  2.353194756515655
  1.5853030268731967
 -0.6713831865960345

julia> lq(A) \ [3,4]
5-element Array{Float64,1}:
  1.7983028303716546
  0.18647369486561782
  2.353194756515656
  1.5853030268731967
 -0.6713831865960342

julia> ldiv!(lq(A), [3.0,4,0,0,0])
5-element Array{Float64,1}:
  1.7983028303716546
  0.18647369486561782
  2.353194756515656
  1.5853030268731967
 -0.6713831865960342

julia> ldiv!(qr(A, Val(true)), [3.0,4,0,0,0])
5-element Array{Float64,1}:
  1.798302830371654
  0.1864736948656177
  2.353194756515655
  1.5853030268731967
 -0.6713831865960345
```
Note the last line — this is **not a new behavior**.  `ldiv!` behaves the same way for pivoted QR when solving underdetermined systems.

Reply:
Thanks. I wasn't aware of that bevavior, I just saw that `\` does that for the caller.